### ISSUE TITLE
Focus ring does not account for radio buttons

### ISSUE THREAD

**Issue by johnevanofski** [WICG/focus-ring#54]
I noticed that the focus-ring polyfill only looks for tab key presses, and not arrow input, which is part of the default behavior for radio buttons. 

I've fixed this locally by adding a few lines to the `onKeyDown` function.

```
function onKeyDown(e) {
    if (e.altKey || e.ctrlKey || e.metaKey)
      return;

    if (
	e.keyCode != 9 // tab
	&& e.keyCode != 37 // left arrow
	&& e.keyCode != 38 // up arrow
	&& e.keyCode != 39 // right arrow
	&& e.keyCode != 40 // down arrow
	)
      return;

    hadKeyboardEvent = true;
  }
```

**Comment by fstorr** [comment]
@johnevanofski Thanks for this—it's just what I needed.

**Comment by robdodson** [comment]
@alice can you think of any reason why we wouldn't want to add this to the library? Otherwise I think we should PR it

**Comment by robdodson** [comment]
@alice friendly ping on this. are you cool with me doing a PR or can you think of a reason we wouldn't want to include this?

**Comment by majornista** [comment]
@robdodson I use the following to evaluate `event.key`, that way the polyfill will work for tablists, list boxes, radio buttons, or a dialog closed with Escape key or keyboard activation of an OK button:

```js
  var NAVIGATION_KEYS = [
    'Tab',
    'ArrowUp',
    'ArrowRight',
    'ArrowDown',
    'ArrowLeft',
    'Home',
    'End',
    'PageUp',
    'PageDown',
    'Enter',
    ' ',
    'Escape',

    /* IE9 and Firefox < 37 */
    'Up',
    'Right',
    'Down',
    'Left',
    'Esc'
  ];

  function onKeyDown(e) {
    if (e.altKey || e.ctrlKey || e.metaKey || NAVIGATION_KEYS.indexOf(event.key) === -1)
      return;

    hadKeyboardEvent = true;
  }
```

**Comment by robdodson** [comment]
@kloots I'm curious what your thoughts are on this? Do you think expanding the keys that trigger keyboard modality to the above set will affect your keyboard shortcut use case?

**Comment by kloots** [comment]
@robdodson I think @johnevanofski is right, we _should_ account for this behavior as browsers do move focus between radio buttons in a group (radio buttons with the same `name`). So, the `onKeyDown` listener should first check if the target of the event is a radio button before responding to an arrow key press.

**Comment by majornista** [comment]
@kloots and @robdodson Browser doesn't natively move focus in response to Home, End, PageUp, PageDown, Enter, Space, Esc, but components often do. Much like radio buttons, in WAI-ARIA design pattern, pressing Home or End on a `tab` in a `tablist` moves focus to first or last `tab`. Closing a dialog using `Esc` key, or by pressing `Enter` or `Space` on a Cancel button, should move focus to the button element that had focus before it was opened and one should expect the visible focus as if focus was set with tab key. Any sort of data grid, with focus and selection, would probably also benefit. In my PR, I don't flag `hadKeyboardEvent` on `document.body`, so Home, End, PageUp, PageDown or Space, should continue to behave as expected when no element has focus.

**Comment by kloots** [comment]
@majornista Right, but those custom widgets also manage their own focus/manipulate focus and as such should apply the `focus-ring` class when they want the focus ring to appear in response to any of those keys. If this polyfill tries to account for custom controls, we'll be right back where we started before issue #33.

**Comment by majornista** [comment]
@kloots and @robdodson Sorry for not being aware of the earlier thread regarding Issue #33 and PR #34. I'm with you @kloots, that the previous way of interpreting any keyboard interaction as a trigger applying `:focus-ring` to the next element to receive focus was problematic. 

I also think that only handling `Tab` and `Shift + Tab` is a bit limiting too. If this is meant as a polyfill for `:focus-ring`, I would kind of expect it to work with custom elements upon which you add `.focus-ring` as if you were adding `:focus-ring`. Why would you bother to use something like postcss-focus-ring otherwise?

**Comment by robdodson** [comment]
> There does seem to be a balance that can be struck, and perhaps it involves aligning keyboard input design patterns with types of controls.

This is a pretty interesting idea.

I wonder if @alice's proposal for [an additional CSS property](https://github.com/WICG/focus-ring/issues/42#issuecomment-309309323) would help.

```css
[role="tab"] {
  /* match when focus is moved via the keyboard explicitly, i.e. tab/shift-tab, arrows, home/end */
  show-focusring: explicit-focus;
}

[role="tab"]:focus-ring {
  outline: none;
  background: lightgreen;
}
```

**Comment by majornista** [comment]
@robdodson I agree that @alice's proposal would probably provide the level of control necessary, but how should this work in the polyfill?

**Comment by robdodson** [comment]
@alice mentioned an interesting idea and I'm curious what you all think:

What if when a user presses tab to move focus from one control to the next, from then on we assume keyboardModality is true, and any focused element will match :focus-ring. If we detect any event from a fine-grained pointing device (i.e. `mousemove`, `touchstart`, `pointerdown`) we set keyboardModality to false. If they hit tab to move focus again, we set keyboardModality back to true.

Would that solve this issue and any other questions around when `:focus-ring` should match? Are there any obvious flaws to this proposal that we've overlooked? cc @bkardell

**Comment by majornista** [comment]
I agree that mouse down should clear focus-ring, but that still won't account for arrow keys, etc. with radio buttons, tablist, grids. or space bar on a checkbox. My thinking was that the polyfill was trying to match browser input modality heuristics for button, radio and checkbox when using :focus-ring on any element.

**Comment by robdodson** [comment]
@alice and I have been chatting about this issue off and on for a week now. Sorry for the slow reply. It's a tough problem! 😸 

A while back [we changed the polyfill](https://github.com/WICG/focus-ring/pull/34) so the only key that triggers keyboard modality is the `Tab` key. We did this after getting feedback that having _any_ key trigger keyboard modality was causing issues for apps that heavily rely on keyboard shortcuts. Those shortcuts are used to move focus around an app (for example, open a drawer of settings and focus it) and mouse/touch users were surprised to see those panels suddenly have focus rings.

That change created the issue we now face, which is that some controls (e.g. `<input type=”radio”>`) allow the user to move around with keys other than `Tab`, and those states should match `:focus-ring`. The solutions presented so far are:

- Tell developers making a custom control to manually apply the `.focus-ring` class in response to key presses.
- OR, [update the polyfill](https://github.com/WICG/focus-ring/pull/57/commits/787428dcce2ef55027489c031ae054feddacbe0f) to be really smart and for every known ARIA role, figure out what the keyboarding heuristic should be.

The first option would kind of work in the polyfill case, but doesn't help when `:focus-ring` is native since there's no way to manually apply `:focus-ring`. You'd have to invent your own class and apply that everywhere. If you’re doing all of that work then `:focus-ring` isn’t offering you much value.

The second option presents big hurdles as well. Native elements like `<input type=”checkbox”>` `<button>`, and `<select>` have built-in keyboard support and focus ring behavior. I don’t think this behavior is specced, but instead the browsers just copy what one another do (and inconsistently I might add). Importantly, there is no pre-existing heuristic that browsers use for ARIA roles (`tab`, `menuitem`, `gridcell`, etc.). If you focus something that’s `role=”tab”` and hit the right arrow key, nothing will happen. For the polyfill (and eventually `:focus-ring`) to adopt this behavior we would need to spec all of these interactions and include them as part of the `:focus-ring` spec.

Both of these have big issues and the more we’ve thought about it, the more it seems like we’re trying to pack too much into `:focus-ring`. So we’d like to propose a third approach.

We’d like to restore the previous behavior where any key triggers keyboard modality. To solve the keyboard shortcut use case where a developer just wants to _manage_ focus on the page and have it start from some new location, we’d like to explore adding [focus navigation start point](https://developers.google.com/web/updates/2016/03/focus-start-point) as a new DOM API. The idea is you would call `focus()` or a similarly named method on a non-interactive control, and it would just make it so screen reader focus moved to that area and if the user happened to press the `Tab` key it would start searching for interactive elements using that new root. Because these elements are non-interactive, they should not match `:focus-ring`. This would also do away with the need to put `tabindex=-1` on non-interactive elements just so you can move focus to them.

This work would not be part of `:focus-ring` but would instead be a new WICG project.

We’d also like to explore [Alice’s suggestion](https://github.com/WICG/focus-ring/issues/42#issuecomment-309309323) of a new CSS property that lets any element indicate when/how it should match `:focus-ring`. This would also be a new WICG spec.

This is a pretty big change we’re proposing so we’d like to hear what others think :)

**Comment by bkardell** [comment]
I'm pro exploring focus navigation starting point but I have to say that as a frequent keyboard user, I'm not sure we have the experience/data to say how that really should play out really well with focus indicators in all cases.  When a drawer or dialog opens, the resulting change really serves as a pretty good implicit indicator of where my focus is, but in several cases I find that when its intent is similar to that of focus navigation starting point I find myself lost - where am I?  It's possible I'm very much in the minority but I don't see universals on this yet and find it frequently disorienting.

**Comment by robdodson** [comment]
> when its intent is similar to that of focus navigation starting point I find myself lost

Can you provide an example of this?

**Comment by bkardell** [comment]
I mean, it's tough without a concrete UI to point to but I've been on sites where it isn't clear what 'main' is and key to main and the only thing that happens is focus disappears and I go "where am I".  if there isn't a pretty clear implication because the area gets a ring or the screen scrolls, I feel kind of lost.  This also happens for me frequently when I close a dialog or a drawer.  While it makes some kind of sense that focus is implied by the sudden appearance of a dialog or drawer, it is frequently less directly clear where focus returns to when the dialog is closed.  Frequently it 'returns' to things that are tabindex=-1 and seem like they'd rather be setting the focus navigation starting point.  idk, in some ways it is very like what we do now, in others not - I'm just a little fuzzy on where it will ultimately wind up/what it will change I guess.

**Comment by robdodson** [comment]
I think the change we're proposing might actually be beneficial for your use case then.

If a developer would like to draw a focus indicator then they can continue to manage focus with the tabindex=-1 trick and that would match `:focus-ring`.

If they decide that the on screen UI change is clear enough that it's implied focus has moved into the new container, then they could use focus start point.

**Comment by kloots** [comment]
@robdodson, @bkardell [my fix for issue #15](https://github.com/WICG/focus-ring/pull/60) ended up requiring me to fix this issue.

**Comment by alice** [comment]
I don't think that fix actually does end up fixing this (although it is a great fix for the other issue): if you focus on the first radio button on https://wicg.github.io/focus-ring/demo/ and then press down, the second radio button doesn't get a focus ring.

What if instead of focusing only on tab, we had a whitelist of keys which trigger focus modality? It seems like arrow keys are a fairly strong signal.

**Comment by majornista** [comment]
@alice You mean something like this: https://github.com/WICG/focus-ring/pull/57 ;)

**Comment by robdodson** [comment]
@kloots what do you think of the proposal in https://github.com/WICG/focus-ring/issues/54#issuecomment-337783170 and the notion of focus navigation start point?

**Comment by kloots** [comment]
@robdodson sorry for my very late reply to this issue. [I've got a PR up that fixes this specific bug](https://github.com/WICG/focus-ring/pull/63). Going to re-read this thread to formulate an opinion on the proposal for this focus navigation start point.

**Comment by robdodson** [comment]
Thanks for sending in the PR :)

I'm going to hold off on merging it based on your feedback on the thread. I'm basically proposing reverting the behavior because I think we're trying to solve two slightly different problems with the same tool.

**Comment by robdodson** [comment]
> That said, I don't follow how it relates to the problem in this specific thread. But still, it's a great idea.

I was suggesting that we simplify `:focus-ring` so that it goes back to matching if _any_ keyboard key was used to move focus. Then we don't need to add special consideration for arrow keys, home/end, pageup/pagedown, etc. We also don't need to account for ARIA patterns.

For Slack's use case, where y'all use keyboard shortcuts to manage focus to non-interactive regions of the page, you would switch to using focus navigation start point. Focus navigation start point would not match `:focus` or `:focus-ring`, so if you use a keyboard shortcut to move focus into a drawer, there wouldn't be a focus ring drawn.

**Comment by srinivasvenigalla** [comment]
I have noticed that when the focus is set to an element currently not in viewport, the focused element does not scroll into view. I have seen this in Chrome, but not tested in other browsers. It would be nice if we can add element.scrollIntoView after setting the focus. https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView#Browser_compatibility

**Comment by robdodson** [comment]
Hey @srinivasvenigalla. There's an [open Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=776368&can=5&q=&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified) for focus() interoperability.

Calling `scrollIntoView` is outside of the scope of this polyfill since we're only trying to simulate the CSS feature `:focus-ring`.

**Comment by kloots** [comment]
@robdodson regarding:

> I was suggesting that we simplify :focus-ring so that it goes back to matching if any keyboard key was used to move focus. Then we don't need to add special consideration for arrow keys, home/end, pageup/pagedown, etc. We also don't need to account for ARIA patterns.

It's still unclear to me _why_ we need to go back to the old behavior.

1. The special consideration for arrow keys is unique to **one** native control type: radio group. So, if we merge my PR we can check that off.

2. I still don't believe we _have_ to account for ARIA patterns as part of this project. We certainly _can_, but I disagree with your earlier concern that the polyfill will be less valuable if it doesn't. The reason being: if you're implementing custom controls for keyboard behavior in your web app, a couple of extra `addClass` and `removeClass` calls is pretty trivial.

Can you explain further why you think the return to the old behavior is necessary given those points?

**Comment by robdodson** [comment]
> The reason being: if you're implementing custom controls for keyboard behavior in your web app, a couple of extra addClass and removeClass calls is pretty trivial.

Since `:focus-ring` is a pseudo-class, a developer would not be able to apply it directly. Most of the ARIA patterns I've seen that require keyboard navigation also suggest that home/end, pageup/pagedown be supported. Drawing the native focus indicator in response to a developer calling `focus()` after one of those keys are pressed in their custom control seems like a useful feature. There are also quirky things in the platform like [`accesskey`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/accesskey) and it'd be nice if `:focus-ring` matched in response to that as well.

It seems like the default behavior one would want is that when the keyboard, or some other coarse-grained pointing device is used to move focus, the newly focused element should match `:focus-ring`. A possible exception to this rule is if the element is `tabindex=-1`. I think that would solve the issue around managing focus to non-interactive regions of the page in response to a keyboard shortcut.

Because it's hard to say with 100% certainty what the developer will always want, I'd like to combine this behavior with Alice's proposed [`show-focus-ring` CSS property](https://github.com/WICG/focus-ring/issues/42#issuecomment-309309323) so developers could tell the system exactly what they want. The tools are then much easier to understand: `:focus-ring` matches in response to keyboard activity, and if it does something surprising/annoying you can override it with `show-focus-ring`.

**Comment by robdodson** [comment]
The actual radio button issue should be fixed in #68. I'll open another thread to discuss adding Alice's show-focus-ring property and we can explore that work on a branch.
